code/go/0chain.net/smartcontract/provider/provider.go

```diff
@@ -1,17 +1,70 @@
 package provider
 
 import (
+	"fmt"
+
+	"0chain.net/core/common"
 	"0chain.net/core/datastore"
 	"0chain.net/smartcontract/stakepool/spenum"
 )
 
 //go:generate msgp -io=false -tests=false -v
 
+type Abstract interface {
+	IsActive(common.Timestamp, common.Timestamp) (bool, string)
+	Kill()
+	IsKilled() bool
+	IsShutDown() bool
+	Id() string
+	Type() spenum.Provider
+	ShutDown()
+}
+
+type Provider struct {
+	ID              string           `json:"id" validate:"hexadecimal,len=64"`
+	LastHealthCheck common.Timestamp `json:"last_health_check"`
+	HasBeenShutDown bool             `json:"is_shut_down"`
+	HasBeenKilled   bool             `json:"is_killed"`
+	ProviderType    spenum.Provider  `json:"provider_type"`
+}
+
 func GetKey(id string) datastore.Key {
 	return "provider:" + id
 }
 
-type Provider struct {
-	ID           string          `json:"id" validate:"hexadecimal,len=64"`
-	ProviderType spenum.Provider `json:"provider_type"`
+func (p *Provider) Id() string {
+	return p.ID
+}
+
+func (p *Provider) IsActive(now, healthCheckPeriod common.Timestamp) (bool, string) {
```

Comment:
can we have status instead with values Active, Inactive, Shutdown, Killed values?

Reply:
Yes, that is what I tried the first time.

On reflection, I thought a single `IsActive` method was simpler and needed less ongoing maintenance. But maybe there is a point I overlooked.

It is easy to get confused about the relationships between (Active, Inactive, Shutdown, Killed). Really all we want to know is if the provider is at least one out of Inactive, shutdown, killed, done a health check recently or anything else (such as just reduced one of the `max max` global settings, so the provider is no longer valid).